File: app/src/main/java/com/kjjejones42/audiobookplayer/display/FileScannerWorker.java
package com.kjjejones42.audiobookplayer.display;

import android.annotation.SuppressLint;
import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.ArraySet;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.kjjejones42.audiobookplayer.AudioBook;
import com.kjjejones42.audiobookplayer.MediaItem;
import com.kjjejones42.audiobookplayer.database.AudiobookDao;
import com.kjjejones42.audiobookplayer.database.AudiobookDatabase;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

public class FileScannerWorker extends Worker {

    private static final String[] authorFields = new String[]{
            MediaStore.Audio.Media.ARTIST,
            MediaStore.Audio.Media.ALBUM_ARTIST,
            MediaStore.Audio.Media.AUTHOR,
            MediaStore.Audio.Media.COMPOSER,
            MediaStore.Audio.Media.WRITER
    };

    public FileScannerWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    private String getFileAuthor(Uri filename) {
        try (Cursor cursor = getApplicationContext().getContentResolver().query(
                MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                authorFields,
                MediaStore.Audio.Media._ID + " = ?",
                new String[]{ filename.getLastPathSegment() },
                null
        )){
            if (cursor == null) return null;
            while (cursor.moveToNext()) {
                for (int i = 0; i < cursor.getColumnCount(); i++) {
                    String result = cursor.getString(i);
                    if (result != null && !result.isEmpty() && !result.equals("<unknown>")) {
                        return result;
                    }
                }
            }
        }
        return null;
    }

    private String findAuthor(Collection<MediaItem> mediaFiles) {
        for (MediaItem item : mediaFiles) {
            String author = getFileAuthor(item.getUri());
            if (author != null) {
                return author;
            }
        }
        return "";
    }

    @SuppressLint("Range")
    private String findImage(String directory) {
        Cursor cursor = getApplicationContext().getContentResolver().query(
                MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                new String[]{ MediaStore.Images.Media.DATA },
                MediaStore.Images.Media.RELATIVE_PATH + " = ?",
                new String[]{ directory },
                null
        );
        String result = null;
        if (cursor != null) {
            if (cursor.moveToNext()) {
                result = cursor.getString(cursor.getColumnIndex(MediaStore.Images.Media.DATA));
            }
            cursor.close();
        }
        return result;
    }

    private AudioBook parseBook(String rel, List<MediaItem> mediaFiles) {
        String directory = new File(rel).getName();
        String imagePath = findImage(rel);
        String author = findAuthor(mediaFiles);
        return new AudioBook(directory, rel, imagePath, mediaFiles, author);
    }

    @SuppressLint("Range")
    private List<AudioBook> getList() {
        List<AudioBook> results = new ArrayList<>();

        Cursor cursor = getApplicationContext().getContentResolver().query(
            MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                new String[]{
                    MediaStore.Audio.Media._ID,
                    MediaStore.Audio.Media.RELATIVE_PATH,
                    MediaStore.Audio.Media.TITLE,
                    MediaStore.Audio.Media.DURATION,
                },
                MediaStore.Audio.Media.IS_AUDIOBOOK + " != 0",
                null,
                null
        );
        if (cursor != null) {
            Map<String, List<MediaItem>> dirs = new HashMap<>();
            while (cursor.moveToNext()) {
                long id = cursor.getLong(0);
                String dir = cursor.getString(1);
                String title = cursor.getString(2);
                int duration = cursor.getInt(3);
                Uri uri = ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, id);
                MediaItem media = new MediaItem(uri, title, duration);
                if (!dirs.containsKey(dir)) {
                    dirs.put(dir, new ArrayList<>());
                }
                Objects.requireNonNull(dirs.get(dir)).add(media);
            }
            cursor.close();
            for (Map.Entry<String, List<MediaItem>> i : dirs.entrySet()) {
                AudioBook result = parseBook(i.getKey(), i.getValue());
                results.add(result);
            }
        }
        return results;
    }

    @NonNull
    @Override
    public Result doWork() {
        try {
            List<AudioBook> results = getList();
            Set<String> ids = results.stream().map(x -> x.baseDir).collect(Collectors.toSet());
            AudiobookDao dao = AudiobookDatabase.getInstance(getApplicationContext()).audiobookDao();
            Set<String> dbIds = new ArraySet<>(dao.getAllBaseDirs());
            dbIds.removeAll(ids);
            for (String dbId : dbIds) {
                dao.delete(dbId);
            }
            dao.insertAll(results);
        } catch (Exception e) {
            e.printStackTrace();
            return Result.failure();
        }
        return Result.success();
    }
}

File: app/src/main/java/com/kjjejones42/audiobookplayer/display/FileScannerWorker.kt
package com.kjjejones42.audiobookplayer.display

import android.annotation.SuppressLint
import android.content.ContentUris
import android.content.Context
import android.net.Uri
import android.provider.MediaStore
import android.util.ArraySet
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.kjjejones42.audiobookplayer.database.AudiobookDatabase.Companion.getInstance
import com.kjjejones42.audiobookplayer.database.models.AudioBook
import com.kjjejones42.audiobookplayer.database.models.AudioBookFile
import com.kjjejones42.audiobookplayer.logError
import java.io.File

class FileScannerWorker(context: Context, workerParams: WorkerParameters) :
    Worker(context, workerParams) {
    private fun getFileAuthor(filename: Uri): String? {
        val uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
        val selection = authorFields
        val where = "${MediaStore.Audio.Media._ID} = ?"
        val args = arrayOf(filename.lastPathSegment)
        applicationContext.contentResolver.query(uri, selection, where, args, null)?.use { cursor ->
            while (cursor.moveToNext()) {
                for (i in 0..<cursor.columnCount) {
                    cursor.getString(i)
                        ?.takeIf { it.isNotEmpty() && (it != "<unknown>")  }
                        ?.let { return it }
                }
            }
        }
        return null
    }

    @SuppressLint("Range")
    private fun findImage(directory: String): String? {
        val uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
        val selection = arrayOf(MediaStore.Images.Media.DATA)
        val where = "${MediaStore.Images.Media.RELATIVE_PATH} = ?"
        val args = arrayOf(directory)
        applicationContext.contentResolver.query(uri, selection, where, args, null)?.use {
            while (it.moveToNext()) {
                val columnIndex = it.getColumnIndex(MediaStore.Images.Media.DATA)
                it.getString(columnIndex)?.let {
                    return it
                }
            }
        }
        return null
    }

    private fun parseBook(rel: String, mediaFiles: List<AudioBookFile>): AudioBook {
        val directory = File(rel).name
        val imagePath = findImage(rel)
        val author = mediaFiles.firstNotNullOfOrNull { getFileAuthor(it.uri) } ?: ""
        return AudioBook(directory, rel, imagePath, mediaFiles, author)
    }

    @SuppressLint("Range")
    private fun queryBooks(): List<AudioBook?> {
        val result = ArrayList<AudioBook>()
        val selection = arrayOf(
            MediaStore.Files.FileColumns._ID,
            MediaStore.Files.FileColumns.RELATIVE_PATH,
            MediaStore.Files.FileColumns.TITLE,
            MediaStore.Files.FileColumns.DISPLAY_NAME,
            MediaStore.Files.FileColumns.DURATION
        )

        val where = "${MediaStore.Files.FileColumns.MIME_TYPE} LIKE ? AND ${MediaStore.Files.FileColumns.RELATIVE_PATH} LIKE ?"
        val selectionArgs = arrayOf("audio/%", "Audiobooks/%")

        for (uri in listOf(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, MediaStore.Audio.Media.INTERNAL_CONTENT_URI)) {
            val cursor = applicationContext.contentResolver.query(uri, selection, where, selectionArgs, null)
            cursor?.use {
                it.apply {
                    val idColumn = getColumnIndex(MediaStore.Files.FileColumns._ID)
                    val dirColumn = getColumnIndex(MediaStore.Files.FileColumns.RELATIVE_PATH)
                    val titleColumn = getColumnIndex(MediaStore.Files.FileColumns.TITLE)
                    val durationColumn = getColumnIndex(MediaStore.Files.FileColumns.DURATION)
                    val fileNameColumn = getColumnIndex(MediaStore.Files.FileColumns.DISPLAY_NAME)

                    val dirs = HashMap<String, MutableList<AudioBookFile>>()
                    while (moveToNext()) {
                        val id = getLong(idColumn)
                        val dir = getString(dirColumn) ?: continue
                        val title = getString(titleColumn)
                        val duration = getInt(durationColumn)
                        val fileName = getString(fileNameColumn)
                        val mediaUri = ContentUris.withAppendedId(uri, id)
                        val media = AudioBookFile(mediaUri, title, fileName, duration.toLong())
                        dirs.computeIfAbsent(dir) { ArrayList() }.add(media)
                    }
                    val uriResult = dirs.map { (key, value) -> parseBook(key, value) }
                    result.addAll(uriResult)
                }
            }
        }
        return result
    }

    override fun doWork(): Result {
        return try {
            val books = queryBooks()
            val ids = books.filterNotNull().mapNotNull { it.baseDir }.toSet()
            val dao = getInstance(applicationContext).audiobookDao()
            val dbIds: MutableSet<String?> = ArraySet(dao.allBaseDirs)
            dbIds.removeAll(ids)
            for (dbId in dbIds) {
                dao.delete(dbId)
            }
            dao.insertAll(books.filterNotNull())
            Result.success()
        } catch (e: Exception) {
            logError(e, "Error scanning files", applicationContext)
            Result.failure()
        }
    }

    companion object {
        private val authorFields by lazy {
            arrayOf(
                MediaStore.Audio.Media.ARTIST,
                MediaStore.Audio.Media.ALBUM_ARTIST,
                MediaStore.Audio.Media.AUTHOR,
                MediaStore.Audio.Media.COMPOSER,
                MediaStore.Audio.Media.WRITER
            )
        }
    }
}

